File: cgma-igm-comparison-ui-app/python-backend/app/cgma_client.py
import time
import logging
import requests
from app.config import config

logger = logging.getLogger(__name__)
MAX_RETRIES = 3


def _with_retry(fn):
    """Retry with exponential backoff."""
    for attempt in range(MAX_RETRIES + 1):
        try:
            return fn()
        except requests.RequestException as e:
            if attempt == MAX_RETRIES:
                raise
            delay = 0.5 * (2 ** attempt)
            logger.warning(f"Retry {attempt + 1}/{MAX_RETRIES} after error: {e}")
            time.sleep(delay)


def get_filepaths(date: str) -> list[dict]:
    """Fetch CGMA document filepaths for a given date."""
    def _call():
        resp = requests.get(
            f"{config.cgma_api_host}/api/documents/filepath",
            params={"date": date},
            headers={"X-API-KEY": config.cgma_api_key, "Accept": "application/json"},
        )
        resp.raise_for_status()
        return resp.json()
    return _with_retry(_call)

File: cgma-igm-comparison-ui-app/backend/src/cgma_client.rs
use reqwest::Client;
use serde::Deserialize;

use crate::retry::with_retry;

const MAX_RETRIES: u32 = 3;

#[derive(Debug, Deserialize, Clone)]
pub struct CgmaDocument {
    pub doc_mrid: String,
    pub filename: String,
    pub filepath: String,
}

pub struct CgmaClient {
    client: Client,
    host: String,
    api_key: String,
}

impl CgmaClient {
    pub fn new(host: String, api_key: String) -> Self {
        Self {
            client: Client::new(),
            host,
            api_key,
        }
    }

    /// Fetch CGMA document filepaths for a given date
    pub async fn get_filepaths(
        &self,
        date: &str, // "2026-03-08"
    ) -> Result<Vec<CgmaDocument>, reqwest::Error> {
        with_retry(MAX_RETRIES, || async {
            self.client
                .get(format!("{}/api/documents/filepath", self.host))
                .query(&[("date", date)])
                .header("X-API-KEY", &self.api_key)
                .header("Accept", "application/json")
                .send()
                .await?
                .error_for_status()?
                .json()
                .await
        })
        .await
    }
}
